src/steps/check-lead-activity.ts

```diff
@@ -0,0 +1,150 @@
+      const tokenResponse = await this.client.getActivityPagingToken(sinceDate);
+      const nextPageToken = tokenResponse.nextPageToken;
+
+      const lead = (await this.client.findLeadByEmail(email)).result[0];
+
+      /* Error when lead is not found */
+      if (!lead) {
+        return this.error('Lead %s was not found', [
+          email,
+        ]);
+      }
+
+      const activityTypes = (await this.client.getActivityTypes()).result;
+      const activityType = activityTypes.find(type => type[isNaN(activityTypeIdOrName) ? 'name' : 'id'] == activityTypeIdOrName);
+
+      /* Error when activity is not found */
+      if (!activityType) {
+        return this.error('Activity with ID or Name %s was not found.', [
+          stepData.activityTypeIdOrName,
+        ]);
+      }
+
+      activityTypeIdOrName = activityType.id;
+
+      const activityResponse = await this.client.getActivities(nextPageToken, lead.id, activityTypeIdOrName);
+      const activities = activityResponse.result;
+
+      /* Fail when when the activity supplied is not found in the lead's logs. */
+      if (!activities) {
+        return this.fail('Activity %s was not found for lead %s within the last %d minute(s)', [
+          stepData.activityTypeIdOrName,
+          email,
+          minutesAgo,
+        ]);
+      }
+
+      /* Expected attributes passed to test step. Translate object/map as array for easier comparison with actual attributes */
+      const expectedAttributes = Object.keys(withAttributes).map((key) => { return { name: key, value: withAttributes[key] }; });
+
+      /* Assert Actual vs Expected attributes and pass if at least one activity matches attributes. Otherwise fail */
+      if (expectedAttributes.length > 0) {
+        let validated = false;
+        for (const activity of activities) {
+          const primaryAttribute = this.getPrimaryAttribute(activityTypes, activity);
+          const actualAttributes = activity.attributes;
+          if (primaryAttribute) {
+            actualAttributes.push(primaryAttribute);
+          }
+
+          if (this.hasMatchingAttributes(actualAttributes, expectedAttributes)) {
+            validated = true;
+            break;
+          }
+        }
+
+        if (validated) {
+          return this.pass('Activity %s was found for lead %s within the last %s minute(s). With expected attributes: \n\n', [
+            stepData.activityTypeIdOrName,
+            email,
+            minutesAgo,
+            JSON.stringify(expectedAttributes, null, 2),
+          ]);
+        }
+
+        return this.fail('Expected attributes of activity %s for lead %s within the last %d minute(s) did not match the actual activity attributes. Actual attributes: \n\n', [
+          stepData.activityTypeIdOrName,
+          email,
+          minutesAgo,
+          JSON.stringify(activities.map(activity => activity.attributes), null, 2),
```

Comment:
- [ ] One last rec. on this: even with the pretty formatting, it's a little difficult to parse the different activities at a glance.

<img width="832" alt="Screen Shot 2019-12-10 at 9 15 59 PM" src="https://user-images.githubusercontent.com/3496491/70593605-8b08b400-1b92-11ea-9370-36aaff62b104.png">

Would recommend removing the two layers of array braces (substituting with two new lines), and saving vertical space by re-collapsing the attributes back to plain object.

```
  ✘ Validate that there should be a Change Data Value activity for marketo lead mla-check@thisisjust.atomatest.com in the last 15 minutes

    Expected attributes of activity Change Data Value for lead mla-check@thisisjust.atomatest.com within the last 15 minute(s) did not match the actual activity attributes. Actual attributes:

{
  "New Value", "Web API",
  "Old Value": null,
  "Reason": "Changed by Smart Campaign Automatest Live Demo.Lead routing API action Change Data Value",
  "Source": "Marketo Flow Action",
  "Campaign": "Automatest Live Demo.Lead routing API"
  "Attribute Name": "Lead Source"
}

{
  "New Value": "Inside Sales",
  "Old Value": null,
  ...
}
```